pkg/csi/service/vanilla/controller.go

```diff
@@ -1709,14 +1724,220 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
+						// In future, we need to change govmomi to throw a NotFound error and catch that instead.
+						errMssgFromPBM := fmt.Sprintf("no pbm profile found with name: %q",
+							scParams.StoragePolicyName)
+						if err.Error() == errMssgFromPBM {
+							errMsg := fmt.Sprintf("Storage policy name %q not found in VC %q",
+								scParams.StoragePolicyName, vcenter.Config.Host)
+							log.Warn(errMsg)
+							continue
+						}
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+							"failed to get policy ID for storage policy name %q. Error: %+v",
+							scParams.StoragePolicyName, err)
+					}
+					log.Infof("Found ID %q for storage policy name %q in vCenter %q", storagePolicyID,
+						scParams.StoragePolicyName, vcenter.Config.Host)
+				}
+
+				var candidateDatastores []*cnsvsphere.DatastoreInfo
+				// Get accessible datastores in the matching topology requirement.
+				if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.TopologyPreferentialDatastores) {
+					vcenter, err := c.managers.VcenterManager.GetVirtualCenter(ctx, vcHost)
+					if err != nil {
+						return nil, faultType, logger.LogNewErrorCodef(log, codes.Internal,
+							"failed to get vCenter. Error: %+v", err)
+					}
+					if topologyRequirement.GetPreferred() != nil {
+						log.Debugf("Using preferred topology")
+						candidateDatastores, err = getCandidateDSInTopologyForFileVolumes(ctx,
+							topologyRequirement.GetPreferred(), vcenter)
+						if err != nil {
+							log.Errorf("Error finding shared datastores using preferred topology: %+v",
+								topologyRequirement.GetPreferred())
+							return nil, csifault.CSIInternalFault, err
+						}
+					}
+					// If there are no shared datastores for the preferred topology requirement, fetch shared
+					// datastores for the requisite topology requirement instead.
+					if len(candidateDatastores) == 0 && topologyRequirement.GetRequisite() != nil {
+						log.Debugf("Using requisite topology")
+						candidateDatastores, err = getCandidateDSInTopologyForFileVolumes(ctx,
+							topologyRequirement.GetRequisite(), vcenter)
+						if err != nil {
+							log.Errorf("Error finding shared datastores using requisite topology: %+v",
+								topologyRequirement.GetRequisite())
+							return nil, csifault.CSIInternalFault, err
+						}
+					}
+				}
+				fsEnabledClusterToDsInfoMap = c.authMgrs[vcHost].GetFsEnabledClusterToDsMap(ctx)
+				for _, datastores := range fsEnabledClusterToDsInfoMap {
+					for _, fsEnabledDatastore := range datastores {
+						for _, ds := range candidateDatastores {
+							if ds.Info.Url == fsEnabledDatastore.Info.Url {
+								fsEnabledCandidateDatastores = append(fsEnabledCandidateDatastores, ds)
+							}
+						}
+					}
+				}
+
+				if scParams.StoragePolicyName != "" {
+					storagePolicyID, err := vcenter.GetStoragePolicyIDByName(ctx, scParams.StoragePolicyName)
```

Comment:
we already obtained policy ID above. We can reuse that value.

Reply:
Done